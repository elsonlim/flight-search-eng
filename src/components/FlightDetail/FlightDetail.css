.flight-details .flight-time {
    font-weight: bold;
    font-size: 1.2em;
}

.flight-details .flight-connection {
    height: 100%;
    padding-bottom: 1em;
}