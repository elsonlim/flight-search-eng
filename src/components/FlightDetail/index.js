import React, { Component } from 'react';
import PropTypes from 'prop-types';
import AirlineLogo from '../AirlineLogo';
import './FlightDetail.css';
import { Grid } from 'semantic-ui-react'

/**
 * A FlightDetail object
 */
class FlightDetail extends Component {
    /**
	 * This method change a number to time(i.e. 2359 to 11:59pm)
	 * @returns {String} that displays time.
	 */
    _numToTime(number) {
        let dayFormat = 'am';
        const mins = ("0" + (number % 100)).slice(-2);
        let hours = Math.floor(number/100%100);
        
        if(hours >= 12) {
            dayFormat = 'pm';
        }

        if(hours > 12) {
            hours -= 12;
        }
        
        hours = ("0" + hours).slice(-2);

        return `${hours}:${mins}${dayFormat}`;
    }

    /**
	 * This method adds a indicator is the flight arrives 
	 * @returns {String} a string indicator of day increment.
	 */
    _getDayIncrement(departure, arrival) {
        if(arrival < departure) {
            return "(+1)"
        }
        return "";
    }

    /**
	 * This method generates a fight detail object
	 * @returns {ReactElement} that show flight details.
	 */
    render () {
        const { flightInfo } = this.props;
        const {arrival, departure} = flightInfo.time;
        return (
            <div className="flight-details" >
                <Grid>
                <Grid.Row verticalAlign={'middle'}>
                    <Grid.Column width="1" />
                    <Grid.Column width="4">
                        <div className="logo">
                            <AirlineLogo airlineLogo={flightInfo.IATA} />
                        </div>
                    </Grid.Column>
                    <Grid.Column width="3">
                        <div className="flight-time">{this._numToTime(departure)}</div>
                        <div className="flight-country">{flightInfo.from}</div>
                    </Grid.Column>
                    <Grid.Column width="4">
                        <div className="flight-connection">
                            <div className="flight-number">{flightInfo.IATA}{flightInfo.flightNo}</div>
                            <hr />
                        </div>                       
                    </Grid.Column>
                    <Grid.Column width="3">
                        <div className="flight-time">
                            {this._numToTime(arrival) + this._getDayIncrement(departure, arrival)}    
                        </div>
                        <div className="flight-country">{flightInfo.to}</div>
                    </Grid.Column>
                </Grid.Row>
                </Grid>
            </div>
        );
    }
}

FlightDetail.propTypes = {
    flightInfo: PropTypes.object.isRequired,
};

export default FlightDetail;