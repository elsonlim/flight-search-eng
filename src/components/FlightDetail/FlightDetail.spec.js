import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import FlightDetail from './index';

enzyme.configure({ adapter: new Adapter() });

describe('FlightDetail', () => {

    let props;
    let wrapper;
    
    beforeEach(() => {
        props = {
            flightInfo: {
                "id": 0,
                "IATA": "DL",
                "flightNo": 90,
                "from": "USA",
                "to": "HKG",
                "price": {
                    "amount": 261,
                    "currency": "USD"
                },
                "time": {
                    "arrival": 609,
                    "departure": 425,
                    "duration": 104
                }
            },
        }
        wrapper = shallow(<FlightDetail {...props} />);
    });

    it('renders without crashing', () => {
		const div = document.createElement('div');
		ReactDOM.render(<FlightDetail {...props} />, div);
		ReactDOM.unmountComponentAtNode(div);
    });
    
    describe('snapshot test', () => {
		test('show airline logo', () => {
			const tree = renderer.create(<FlightDetail {...props} />).toJSON();
			expect(tree).toMatchSnapshot();
		});
    });
    
    describe('_numToTime', () => {
        test('convert 2 to 00:02am', () => {
            expect(wrapper.instance()._numToTime(2)).toEqual('00:02am');
        });

        test('convert 102 to 01:02am', () => {
            expect(wrapper.instance()._numToTime(102)).toEqual('01:02am');
        });

        test('convert 1234 to 00:34am', () => {
            expect(wrapper.instance()._numToTime(1234)).toEqual('12:34pm');
        });

        test('convert 2359 to 11:59pm', () => {
            expect(wrapper.instance()._numToTime(2359)).toEqual('11:59pm');
        });
    });

    describe('_getDayIncrement', () => {
        test('returns empty string if depature time before arrival', () => {
            expect(wrapper.instance()._getDayIncrement(1234, 2234)).toEqual('');
        });

        test('returns +1 string if depature time before arrival', () => {
            expect(wrapper.instance()._getDayIncrement(2359, 1234)).toEqual('(+1)');
        });
    });
});
