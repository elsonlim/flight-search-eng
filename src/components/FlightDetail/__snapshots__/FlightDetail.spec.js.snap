// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlightDetail snapshot test show airline logo 1`] = `
<div
  className="flight-details"
>
  <div
    className="logo"
  >
    <div>
      <img
        className="ui tiny rounded image"
        src="/airlineLogo/dl.svg"
      />
    </div>
  </div>
  <div
    className="details"
  >
    <div>
      DL
      -
      90
    </div>
    <div>
      USA
       &gt; 
      HKG
    </div>
    <div>
      Depart: 
      04:25am
    </div>
    <div>
      Arrival: 
      06:09am
    </div>
  </div>
</div>
`;
